dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+            inside the object of interest and 0's along the background).
+            If None, the behaviour is equivalent to mmask=ones_like(static)
+        """
+
+        if smask is None:
+            smask = np.ones_like(static)
+        if mmask is None:
+            mmask = np.ones_like(moving)
+
+        self.smin = np.min(static[smask!=0])
+        self.smax = np.max(static[smask!=0])
+        self.mmin = np.min(moving[mmask!=0])
+        self.mmax = np.max(moving[mmask!=0])
+
+        self.sdelta = (self.smax - self.smin)/(self.nbins - 2 * self.padding)
+        self.mdelta = (self.mmax - self.mmin)/(self.nbins - 2 * self.padding)
+        self.smin = self.smin/self.sdelta - self.padding
+        self.mmin = self.mmin/self.mdelta - self.padding
+
+        self.joint_grad = None
+        self.metric_grad = None
+        self.metric_val = 0
+        self.joint = np.ndarray(shape = (self.nbins, self.nbins))
+        self.smarginal = np.ndarray(shape = (self.nbins,), dtype = np.float64)
+        self.mmarginal = np.ndarray(shape = (self.nbins,), dtype = np.float64)
+
+
+    def bin_normalize_static(self, x):
+        r""" Normalizes intensity x to the range covered by the static histogram
+
+        If the input intensity is in [self.smin, self.smax] then the normalized
+        intensity will be in [self.padding, self.nbins - self.padding]
+
+        Parameters
+        ----------
+        x : float
+            the intensity to be normalized
+
+        Returns
+        -------
+        xnorm : float
+            normalized intensity to the range covered by the static histogram
+        """
+        return _bin_normalize(x, self.smin, self.sdelta)
+
+
+    def bin_normalize_moving(self, x):
+        r""" Normalizes intensity x to the range covered by the moving histogram
+
+        If the input intensity is in [self.mmin, self.mmax] then the normalized
+        intensity will be in [self.padding, self.nbins - self.padding]
+
+        Parameters
+        ----------
+        x : float
+            the intensity to be normalized
+
+        Returns
+        -------
+        xnorm : float
+            normalized intensity to the range covered by the moving histogram
+        """
+        return _bin_normalize(x, self.mmin, self.mdelta)
+
+
+    def bin_index(self, xnorm):
+        r""" Bin index associated to the given normalized intensity
+
+        The return value is an integer in [padding, nbins - 1 - padding]
+
+        Parameters
+        ----------
+        xnorm : float
+            intensity value normalized to the range covered by the histogram
+
+        Returns
+        -------
+        bin : int
+            the bin index associated to the given normalized intensity
+        """
+        return _bin_index(xnorm, self.nbins, self.padding)
+
+
+    def update_pdfs_dense(self, static, moving, smask=None, mmask=None):
+        r''' Computes the Joint Probability Density Function of two images
```

Comment:
Comment on where the JPDF goes?  (Into `self.smarginal, self.mmarginal` I guess).